### PR TITLE
feat: key safety and restoration

### PR DESCRIPTION
**Pull request by mkazlauskas**
# Context

The current version of the `InMemoryKeyManager` was designed primarily for testing, with the intent to iterate it into a secure, production-ready implementation.

We're currently holding private key in memory with our `InMemoryKeyManager`. Stealing it with access to someone's laptop would be fairly quick and easy by using Chrome devtools memory heap snapshot, or maybe even just running a script in console. We should probably only decrypt private key when it needs to sign a transaction.

# Proposed Solution

- Implement [EmIP-003](https://github.com/Emurgo/EmIPs/blob/master/specs/emip-003.md) encryption
  - Add tests for Yoroi compatibility:
    - Decrypting Yoroi-encrypted root private key
    - Restoring root private key from Yoroi-generated 15 word mnemonic
- (BREAKING) Refactor key agent to not keep keys in-memory and decrypt on-demand. **Restoring keys from mnemonic will result in a new private key, so any testing/development wallets will be recreated**.
  - Fix an issue that appeared after refactoring some KeyManager functions to be async (`sharedDistinctBlock` -> `distinctBlock` and `sharedDistinctEpoch`->`distinctEpoch`)
  - Implement and export `KeyManagement.cachedGetPassword` to cache password for `InMemoryKeyAgent`. This is needed for any bearable user experience, as pretty much every operation requires decrypting private key.
  - Create custom KeyAgent error type

# Important Changes Introduced

- Add dependencies to the following npm packages:
  - [pbkdf2](https://www.npmjs.com/package/pbkdf2)
  - [chacha](https://www.npmjs.com/package/chacha)
  - [get-random-values](https://www.npmjs.com/package/get-random-values)
- (BREAKING) Rename `KeyManager` to `KeyAgent` and refactor it to support async operations (planning for hardware wallet support)
- New types in `core` package: `Bip32PrivateKey`, `Bip32PrivateKey`, `Signatures`
- Extend `KeyAgent` interface adding a `serializableData` field that will be used for storage. Make it generic enough to be able to implement generic `KeyAgent` serialization in the future.
- Add `KeyAgentBase` abstract class that specific implementations can extend  to reuse some functionality.

### CODE REVIEW

File: packages/wallet/src/KeyManagement/emip3.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
+// Review: a separate package for emip3 might be a good idea because it can be useful as a standalone package.
```
Comment (rhyslbw) [inline packages/wallet/src/KeyManagement/emip3.ts:1]: Agree. We can source it at the top level of this SDK for now

File: packages/wallet/src/KeyManagement/InMemoryKeyAgent.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    return { publicKey, signature };
+  }
+
+  async derivePublicKey({ index, type }: AccountKeyDerivationPath): Promise<Cardano.Ed25519PublicKey> {
+    const accountPrivateKey = await this.#deriveAccountPrivateKey();
+    const cslPublicKey = accountPrivateKey.derive(type).derive(index).to_public().to_raw_key();
+    return Cardano.Ed25519PublicKey(Buffer.from(cslPublicKey.as_bytes()).toString('hex'));
+  }
+
+  // Review: assuming Bip39 entropy is the actual key
```
Comment (rhyslbw) [inline packages/wallet/src/KeyManagement/InMemoryKeyAgent.ts:103]: :+1:

File: packages/wallet/test/KeyManagement/InMemoryKeyManager.test.ts [hunk 3]
```diff
@@ -1,28 +1,26 @@
-import { CSL, Cardano } from '@cardano-sdk/core';
+import { Cardano } from '@cardano-sdk/core';
```
Comment (rhyslbw) [inline packages/wallet/test/KeyManagement/InMemoryKeyManager.test.ts:1]: Looks like you have this noted given the temporary export as the old name, but just a reminder to rename to `InMemoryKeyAgent.test.ts`

File: packages/wallet/src/KeyManagement/types.ts [hunk 4]
```diff
@@ -13,32 +40,51 @@ export enum AddressType {
...
+  encryptedRootPrivateKeyBytes: number[];
+}
+
+export interface SerializableLedgerKeyAgentData extends SerializableKeyAgentDataBase {
+  __typename: KeyAgentType.Ledger;
+}
+
+export type SerializableKeyAgentData = SerializableInMemoryKeyAgentData | SerializableLedgerKeyAgentData;
+
+// TODO: utility to cache password for specified duration
```
Comment (rhyslbw) [inline packages/wallet/src/KeyManagement/types.ts:78]: Would be good to have a doc block for this function. When I think of the behaviour of authentication, I expect the return value to be a boolean, so maybe this needs a different name.

File: packages/wallet/src/KeyManagement/emip3.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (rhyslbw) [reply packages/wallet/src/KeyManagement/emip3.ts:1]: Logged

File: packages/wallet/src/KeyManagement/types.ts (hunk 4 above)
Comment (mkazlauskas) [reply packages/wallet/src/KeyManagement/types.ts:78]: Renamed it to `getPassword`

File: packages/wallet/src/KeyManagement/types.ts [hunk 5]
```diff
@@ -13,32 +36,73 @@ export enum AddressType {
...
+  networkId: Cardano.NetworkId;
+  accountIndex: number;
+}
+
+export interface SerializableInMemoryKeyAgentData extends SerializableKeyAgentDataBase {
+  __typename: KeyAgentType.InMemory;
+  encryptedRootPrivateKeyBytes: number[];
+}
+
+export interface SerializableLedgerKeyAgentData extends SerializableKeyAgentDataBase {
```
Comment (rhyslbw) [inline packages/wallet/src/KeyManagement/types.ts:68]: Maybe drop this and all other Ledger types out until we have a Ledger implementation

File: packages/wallet/src/SingleAddressWallet.ts [hunk 6]
```diff
@@ -233,5 +233,22 @@ export class SingleAddressWallet implements Wallet {
     this.protocolParameters$.complete();
     this.genesisParameters$.complete();
     this.#tip$.complete();
+    this.addresses$.complete();
+  }
+
+  #initializeAddress(existingAddress?: GroupedAddress): Observable<GroupedAddress[]> {
```
Comment (rhyslbw) [inline packages/wallet/src/SingleAddressWallet.ts:239]: This is good for a single-address wallet but doesn't align with multi-address wallet semantics. I think it's fine to just refactor later given it's a private method though

File: packages/wallet/test/KeyManagement/InMemoryKeyAgent.test.ts [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+  test('__typename', () => {
+    expect(typeof keyAgent.__typename).toBe('string');
+  });
+
+  test('accountIndex', () => {
+    expect(typeof keyAgent.accountIndex).toBe('number');
+  });
+
+  test('serializableData', () => {
```
Comment (rhyslbw) [inline packages/wallet/test/KeyManagement/InMemoryKeyAgent.test.ts:32]: A test that demonstrates serializability could be to convert to JSON and back

File: packages/wallet/src/KeyManagement/types.ts (hunk 4 above)
Comment (rhyslbw) [reply packages/wallet/src/KeyManagement/types.ts:74]: The todo can be removed now right?

File: packages/wallet/test/KeyManagement/cachedGetPassword.test.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
+import { KeyManagement } from '../../src';
+
+jest.useFakeTimers();
+
+describe('cachedGetPassword', () => {
+  it('caches password for specified duration and ignores cache when "noCache=true"', async () => {
```
Comment (rhyslbw) [inline packages/wallet/test/KeyManagement/cachedGetPassword.test.ts:6]: Let's break this into two distinct tests for better readability

**Review by rhyslbw — CHANGES_REQUESTED** [review]

**Review by rhyslbw — APPROVED** [review]
Excellent work @mkazlauskas. Great job with the PR description, and conventional commits standards application. :rocket:

**Pull request merged**